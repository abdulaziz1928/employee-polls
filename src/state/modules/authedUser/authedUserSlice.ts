import { PayloadAction, createSlice } from "@reduxjs/toolkit";

const initialState: string | null = null;

const authedUserSlice = createSlice({
  name: "authedUser",
  initialState: initialState as string | null,
  reducers: {
    setAuthedUser: (_, action: PayloadAction<string>) => action.payload,
    logoutAuthedUser: () => null,
  },
});

export const { setAuthedUser, logoutAuthedUser } = authedUserSlice.actions;

export default authedUserSlice.reducer;
