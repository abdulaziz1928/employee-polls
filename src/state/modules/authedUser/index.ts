import reducer from "./authedUserSlice";
export * from "./authedUserSlice";
export default reducer;
