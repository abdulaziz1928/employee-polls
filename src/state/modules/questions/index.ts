import reducer from "./questionsSlice";
export * from "./questionsSlice";
export * from "./models/question";
export default reducer;
