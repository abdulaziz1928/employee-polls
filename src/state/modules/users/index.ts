import reducer from "./usersSlice";
export * from "./usersSlice";
export * from "./models/user";
export default reducer;
