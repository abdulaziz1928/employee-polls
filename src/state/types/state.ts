import LoadingStatus from "./loading_status";

export default interface IState<T> {
  entities: T;
  loading: LoadingStatus;
  error?: string;
}
