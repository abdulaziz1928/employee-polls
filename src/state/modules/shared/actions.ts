import { Dispatch } from "@reduxjs/toolkit";
import { setAuthedUser } from "../authedUser";
import { getInitialData } from "../../apis/api";
import { receiveQuestions } from "../questions";
import { receiveUsers } from "../users";

const AUTHED_USER_ID = "tylermcginnis";

export function handleInitialData() {
  return (dispatch: Dispatch) => {
    return getInitialData().then(({ users, questions }) => {
      dispatch(receiveUsers(users));
      dispatch(receiveQuestions(questions));
      dispatch(setAuthedUser(AUTHED_USER_ID));
    });
  };
}
