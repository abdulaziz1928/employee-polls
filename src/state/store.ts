import authedUser from "./modules/authedUser";
import questions from "./modules/questions";
import users from "./modules/users";
import logger from "./middlewares/logger";
import prefrences from "./modules/prefrences/prefrencesSlice";
import { combineReducers, configureStore } from "@reduxjs/toolkit";

const reducers = { authedUser, users, questions, prefrences };

const rootReducer = combineReducers(reducers);

const middleWares = [logger];

export const store = configureStore({
  reducer: rootReducer,
  middleware: (gDM) => gDM().concat(...middleWares),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
