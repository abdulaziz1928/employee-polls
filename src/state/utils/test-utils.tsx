import React, { PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import type { PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { setupStore, AppStore, RootState } from "../store";
import { MemoryRouter } from "react-router-dom";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderWithProvidersAndRouter(
  ui: React.ReactElement,

  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
  route: string = "/"
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return (
      <Provider store={store}>
        <MemoryRouter initialEntries={[route]}>{children}</MemoryRouter>
      </Provider>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
