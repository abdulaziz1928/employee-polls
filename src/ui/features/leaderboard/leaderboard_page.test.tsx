import LeaderboardPage from "./leaderboard_page";
import { renderWithProviders } from "../../../state/utils/test-utils";
import { setAuthedUser } from "../../../state/modules/authedUser";
import { handleInitialData } from "../../../state/modules/shared/actions";
import { AppStore, setupStore } from "../../../state/store";
import { User } from "../../../state/modules/users";
import { screen } from "@testing-library/react";
import LoadingStatus from "../../../state/types/loading_status";
let store: AppStore;

const users: Record<string, User> = {
  sarahedo: {
    id: "sarahedo",
    password: "password123",
    name: "Sarah Edo",
    avatarURL: "",
    answers: {
      "8xf0y6ziyjabvozdd253nd": "optionOne",
      "6ni6ok3ym7mf1p33lnez": "optionOne",
      am8ehyc8byjqgar0jgpub9: "optionTwo",
      loxhs1bqm25b708cmbf3g: "optionTwo",
    },
    questions: ["8xf0y6ziyjabvozdd253nd", "am8ehyc8byjqgar0jgpub9"],
  },
  tylermcginnis: {
    id: "tylermcginnis",
    password: "abc321",
    name: "Tyler McGinnis",
    avatarURL: "",
    answers: {
      vthrdm985a262al8qx3do: "optionOne",
      xj352vofupe1dqz9emx13r: "optionTwo",
    },
    questions: ["loxhs1bqm25b708cmbf3g", "vthrdm985a262al8qx3do"],
  },
};

describe("LeaderboardPage", () => {
  beforeAll(async () => {
    store = setupStore();
    await store.dispatch(handleInitialData());
    store.dispatch(setAuthedUser("tylermcginnis"));
  });

  it("should render the component", () => {
    const { container } = renderWithProviders(<LeaderboardPage />);
    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();
  });
  it("should render the leaderboard table with users", () => {
    const { container } = renderWithProviders(<LeaderboardPage />, {
      preloadedState: {
        users: { entities: users, loading: LoadingStatus.succeeded },
      },
    });
    const leaderboardBody = screen.getByTestId("leaderboard-table-body");
    expect(leaderboardBody).toHaveTextContent(/Sarah Edo/i);
    expect(leaderboardBody).toHaveTextContent(/tylermcginnis/i);
    expect(container).toMatchSnapshot();
  });
});
