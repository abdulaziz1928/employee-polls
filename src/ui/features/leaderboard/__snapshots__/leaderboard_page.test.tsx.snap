// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeaderboardPage should render the component 1`] = `
<div>
  <span
    class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-indeterminate css-eglki6-MuiLinearProgress-root"
    role="progressbar"
  >
    <span
      class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate css-5ir5xx-MuiLinearProgress-bar1"
    />
    <span
      class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar2Indeterminate css-1r8wrcl-MuiLinearProgress-bar2"
    />
  </span>
</div>
`;

exports[`LeaderboardPage should render the leaderboard table with users 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthMd css-1poorr9-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-0"
    >
      <h1
        class="MuiTypography-root MuiTypography-h3 MuiTypography-alignCenter css-47suzj-MuiTypography-root"
      >
        Leaderboard
      </h1>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-pazu8s-MuiPaper-root-MuiTableContainer-root"
    >
      <div
        class="MuiBox-root css-1gk6p0h"
      >
        <table
          class="MuiTable-root css-rqglhn-MuiTable-root"
        >
          <thead
            class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head css-jkgoj5-MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                colspan="2"
                scope="col"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-jb1ygz-MuiTypography-root"
                >
                  Users
                </p>
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
                scope="col"
              >
                 
                <p
                  class="MuiTypography-root MuiTypography-body1 css-jb1ygz-MuiTypography-root"
                >
                  Answers
                </p>
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
                scope="col"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-jb1ygz-MuiTypography-root"
                >
                  Created
                </p>
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
                scope="col"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-jb1ygz-MuiTypography-root"
                >
                  Score
                </p>
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
            data-testid="leaderboard-table-body"
          >
            <tr
              class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                colspan="2"
              >
                <div
                  class="MuiBox-root css-1i27l4i"
                >
                  <div
                    class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-smk8f9-MuiAvatar-root"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
                      data-testid="PersonIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                      />
                    </svg>
                  </div>
                  <div
                    class="MuiBox-root css-cgq59l"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                    >
                      Sarah Edo
                    </p>
                    <p
                      class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
                    >
                      sarahedo
                    </p>
                  </div>
                </div>
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
              >
                4
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
              >
                2
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
              >
                6
              </td>
            </tr>
            <tr
              class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                colspan="2"
              >
                <div
                  class="MuiBox-root css-1i27l4i"
                >
                  <div
                    class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-smk8f9-MuiAvatar-root"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
                      data-testid="PersonIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                      />
                    </svg>
                  </div>
                  <div
                    class="MuiBox-root css-cgq59l"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                    >
                      Tyler McGinnis
                    </p>
                    <p
                      class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
                    >
                      tylermcginnis
                    </p>
                  </div>
                </div>
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
              >
                2
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
              >
                2
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
              >
                4
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
